src/compiler/crystal/macros/methods.cr

```diff
@@ -1905,7 +1905,13 @@ module Crystal
     end
 
     def self.has_constant?(type, name)
-      BoolLiteral.new(type.types.has_key?(name))
+      return BoolLiteral.new(true) if type.types.has_key?(name)
+
+      name.split "::" do |part|
+        return BoolLiteral.new(false) unless (type = type.types[part]?)
+      end
```

Comment:
Because a global path in any context always refers to the top-level namespace, in theory `@type.has_constant?("::Int32")` should always return true regardless of what `@type` is. This implementation doesn't do that.